ui/jobs/jobs.ts

```diff
@@ -53,39 +58,118 @@ const Options = {
+  init: boolean;
+  o: Partial<{
+    pullCountdown: TimerBar;
+    rightBuffsContainer: HTMLDivElement;
+    leftBuffsContainer: HTMLDivElement;
+    rightBuffsList: WidgetList;
+    leftBuffsList: WidgetList;
+    cpContainer: HTMLDivElement;
+    cpBar: ResourceBar;
+    gpContainer: HTMLDivElement;
+    gpBar: ResourceBar;
+    healthContainer: HTMLDivElement;
+    healthBar: ResourceBar;
+    manaContainer: HTMLDivElement;
+    manaBar: ResourceBar;
+    mpTickContainer: HTMLDivElement;
+    mpTicker: TimerBar;
+  }>;
+
+  // parameters
+  me?: string;
+  level: number;
+  job: Job;
+  hp: number;
+  maxHP: number;
+  currentShield: number;
+  mp: number;
+  prevMP: number;
+  maxMP: number;
+  cp: number;
+  maxCP: number;
+  gp: number;
+  maxGP: number;
+  umbralStacks: number;
+  skillSpeed: number;
+  spellSpeed: number;
+
+  // battle variables
+  distance: number;
+  inCombat: boolean;
+  combo?: ComboTracker;
+  regexes: RegexesHolder;
+  speedBuffs: {
+    presenceOfMind: number;
+    shifu: number;
+    huton: number;
+    lightningStacks: number;
+    paeonStacks: number;
+    museStacks: number;
+    circleOfPower: number;
+  };
+
+  dotTarget: string[];
+  trackedDoTs: string[];
+  comboFuncs: ((skill: string | null) => void)[];
+  // hard to type this when we don't know what job is now
+  jobFuncs: ((jobDetail: never) => void)[];
+  changeZoneFuncs: EventMap['ChangeZone'][];
+  updateDotTimerFuncs: (() => void)[];
+  gainEffectFuncMap: Record<string, JobFunc>;
+  mobGainEffectFromYouFuncMap: Record<string, JobFunc>;
+  mobLoseEffectFromYouFuncMap: Record<string, JobFunc>;
+  loseEffectFuncMap: Record<string, JobFunc>;
+  statChangeFuncMap: Partial<Record<Job, () => void>>;
+  abilityFuncMap: Record<string, JobFunc>;
+  contentType: number;
+  isPVPZone: boolean;
+  crafting: boolean;
+  lastAttackedDotTarget?: string;
+  foodBuffTimer: number | undefined
+  foodBuffExpiresTimeMs: number;
+  gpAlarmReady: boolean;
+  gpPotion: boolean;
+  buffTracker?: BuffTracker;
+
+  constructor(options: IJobsOptions) {
     this.options = options;
     this.init = false;
-    this.me = null;
     this.o = {};
-    this.casting = {};
-    this.job = '';
+
+    this.me = undefined;
+    this.level = 0;
+    this.job = 'NONE';
     this.hp = 0;
     this.maxHP = 0;
     this.currentShield = 0;
     this.mp = 0;
     this.prevMP = 0;
     this.maxMP = 0;
-    this.level = 0;
-    this.distance = -1;
-    this.whiteMana = -1;
-    this.blackMana = -1;
-    this.oath = -1;
+    this.cp = 0;
+    this.maxCP = 0;
+    this.gp = 0;
+    this.maxGP = 0;
     this.umbralStacks = 0;
-    this.inCombat = false;
-    this.combo = null;
-    this.comboTimer = null;
-    this.regexes = new RegexesHolder(this.options.ParserLanguage);
-
     this.skillSpeed = 0;
     this.spellSpeed = 0;
 
+    this.distance = -1;
+    this.inCombat = false;
+    this.combo = undefined;
+    this.regexes = new RegexesHolder(this.options.ParserLanguage);
```

Comment:
I wonder if it would be better to refactor this (ahead of time? in it's own PR?) to only construct `this.regexes` if there's a language and a player name), so that `this.regexes` is the optional piece (and you could early out in log testing if it doesn't exist) rather than having every field on `this.regexes` be independently optional.

Reply:
(*￣3￣)╭ Sounds good, and I think it is better to put in a new PR.